assignment6/CarRearranger.java

```diff
@@ -0,0 +1,160 @@
+package assignment6;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CarRearranger {
+
+    private int[] currentOrder;
+    private int[] destOrder;
+    private int[] placesByCarIndex;
+    private final int EMPTY = 0;
+
+    public CarRearranger(Integer[] sourceOrder, Integer[] destOrder) {
+        validateParameters(sourceOrder, destOrder);
+        init(sourceOrder, destOrder);
+    }
+
+    /**
+     * Validates input data.
+     */
+    private void validateParameters(Integer[] sourceOrder, Integer[] destOrder) {
+        if (sourceOrder == null || destOrder == null) {
+            throw new IllegalArgumentException("Arguments cannot be null.");
+        }
+        if (sourceOrder.length != destOrder.length) {
+            throw new IllegalArgumentException("Source and destination orders have not equal length.");
+        }
+        if (!validateOrder(sourceOrder) || !validateOrder(destOrder)) {
+            throw new IllegalArgumentException("Source or destination order is invalid.");
+        }
+    }
+
+    /**
+     * Checks if array contains numbers from 0 to N-1 and only once.
+     */
+    private boolean validateOrder(Integer[] order) {
+        boolean[] checkExistence = new boolean[order.length];
+        for (int item : order) {
+            if (item < 0 || item >= order.length || checkExistence[item]) {
+                return false;
+            }
+            checkExistence[item] = true;
+        }
+        return true;
+    }
+
+    /**
+     * Initializes data to work with
+     * so that algorithm will not change input data.
+     */
+    private void init(Integer[] sourceOrder, Integer destOrder[]) {
+        this.destOrder = new int[destOrder.length];
+        this.placesByCarIndex = new int[destOrder.length];
+        this.currentOrder = new int[destOrder.length];
+        for (int i = 0; i < sourceOrder.length; i++) {
+            this.placesByCarIndex[sourceOrder[i]] = i;
+            this.currentOrder[i] = sourceOrder[i];
+            this.destOrder[i] = destOrder[i];
+        }
+    }
+
+    public static class Move {
+
+        private int from;
+        private int to;
+
+        public Move(int from, int to) {
+            this.from = from;
+            this.to = to;
+        }
+
+        public int getFrom() {
+            return from;
+        }
+
+        public int getTo() {
+            return to;
+        }
+    }
+
+    /**
+     * Takes O(n) operations.
+     */
+    public List<Move> rearrangeCars() {

```

Comment:
In this case there is no difference between map and array since our keys are always 0, 1, ..., N-1. Really this algorithm has the same logic as yours :) The difference is that I don't take the first car from the map but I iterate over the array to find the first wrong placed car.